packages/eslint-plugin/tests/rules/no-floating-promises.test.ts

```diff
@@ -733,6 +733,41 @@ promise().then(() => {});
         },
       ],
     },
+    {
+      code: `
+interface SafePromise<T> extends Promise<T> {
+  brand: 'safe';
+}
+
+declare const createSafePromise: () => SafePromise<string>;
+createSafePromise();
+      `,
+      options: [
+        {
+          allowForKnownSafePromises: [{ from: 'file', name: 'SafePromise' }],
+          checkThenables: true,
+        },
+      ],
+    },
+    {
+      code: `
+declare const createPromise: () => PromiseLike<number>;
+createPromise();
+      `,
+      options: [{ checkThenables: false }],
+    },
+    {
+      code: `
+interface MyThenable {
+  then(onFulfilled: () => void, onRejected: () => void): MyThenable;
+}
+
+declare function createMyThenable(): MyThenable;
+
+createMyThenable();
+      `,
+      options: [{ checkThenables: false }],
+    },
   ],
 
   invalid: [
```

Comment:
(testing) Looks like there aren't currently any invalid test cases with `checkThenables: false` (since for now the default is `checkThenables: true`). I'm thinking we should add:

- Some invalid tests to ensure that ordinary promises still will flag with the `checkThenables: false` option enabled).
- Some invalid tests around whatever the correct behavior with the `allowForKnownSafePromises` also enabled is.... (also - does allowForKnownSafePromises have any point at all with `checkThenables: false`? Won't the implementation just see branded promises as not the global promise type, and therefore not checked anyway?)

Reply:
> ordinary promises still will flag

👍 added, good call

> does allowForKnownSafePromises have any point at all with `checkThenables: false`?

Ooh yes, added a couple `class MyPromise<T> extends Promise<T> { / * ... */ }` tests. They failed until I switched from `isSymbolFromDefaultLibrary` (one shot) to `isBuiltinSymbolLike` (recursive checks of base types) for the built-in `Promise` check. Nice!